platform/ios/src/MGLMapView.mm

```diff
@@ -4513,7 +4513,26 @@ - (void)notifyMapChange:(mbgl::MapChange)change
                 || self.userTrackingMode == MGLUserTrackingModeNone
                 || self.userTrackingState != MGLUserTrackingStateChanged)
             {
-                [self deselectAnnotation:self.selectedAnnotation animated:NO];
+                // Deselect annotation if it lies outside the viewport
```

Comment:
I agree. Ideally `visibleAnnotations` would be a set, but unfortunately the way we compute it (using feature querying) means it’s always going to be <var>O</var>(<var>n</var>) regardless. (Not <var>O</var>(<var>n</var>²), because we’re only checking whether one particular annotation is visible.)